crates/chia-sdk-driver/src/layers/p2_controller_puzzle_layer.rs

```diff
@@ -0,0 +1,168 @@
+pub struct P2ControllerPuzzleArgs {
+    pub controller_puzzle_hash: Bytes32,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, ToClvm, FromClvm)]
+#[clvm(solution)]
+pub struct P2ControllerPuzzleSolution<P, S> {
+    pub delegated_puzzle: P,
+    pub delegated_solution: S,
+}
+
+impl P2ControllerPuzzleLayer {
+    pub fn new(controller_puzzle_hash: Bytes32) -> Self {
+        Self {
+            controller_puzzle_hash,
+        }
+    }
+
+    pub fn spend(
+        &self,
+        ctx: &mut SpendContext,
+        delegated_spend: Spend,
+    ) -> Result<Spend, DriverError> {
+        self.construct_spend(
+            ctx,
+            P2ControllerPuzzleSolution {
+                delegated_puzzle: delegated_spend.puzzle,
+                delegated_solution: delegated_spend.solution,
+            },
+        )
+    }
+}
+
+impl Layer for P2ControllerPuzzleLayer {
+    type Solution = P2ControllerPuzzleSolution<NodePtr, NodePtr>;
+
+    fn construct_puzzle(&self, ctx: &mut SpendContext) -> Result<NodePtr, DriverError> {
+        let curried = CurriedProgram {
+            program: ctx.p2_controller_puzzle_puzzle()?,
+            args: P2ControllerPuzzleArgs {
+                controller_puzzle_hash: self.controller_puzzle_hash,
+            },
+        };
+        ctx.alloc(&curried)
+    }
+
+    fn construct_solution(
+        &self,
+        ctx: &mut SpendContext,
+        solution: Self::Solution,
+    ) -> Result<NodePtr, DriverError> {
+        ctx.alloc(&solution)
+    }
+
+    fn parse_puzzle(allocator: &Allocator, puzzle: Puzzle) -> Result<Option<Self>, DriverError> {
+        let Some(puzzle) = puzzle.as_curried() else {
+            return Ok(None);
+        };
+
+        if puzzle.mod_hash != P2_CONTROLLER_PUZZLE_PUZZLE_HASH {
+            return Ok(None);
+        }
+
+        let args = P2ControllerPuzzleArgs::from_clvm(allocator, puzzle.args)?;
+
+        Ok(Some(Self {
+            controller_puzzle_hash: args.controller_puzzle_hash,
+        }))
+    }
+
+    fn parse_solution(
+        allocator: &Allocator,
+        solution: NodePtr,
+    ) -> Result<Self::Solution, DriverError> {
+        Ok(Self::Solution::from_clvm(allocator, solution)?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::assert_puzzle_hash;
+
+    use super::*;
+    use chia_protocol::{Bytes, CoinSpend};
+    use chia_sdk_test::Simulator;
+    use chia_sdk_types::Conditions;
+    use clvm_traits::clvm_quote;
+    use clvmr::serde::node_from_bytes;
+
+    #[test]
+    fn test_puzzle_hash() -> anyhow::Result<()> {
+        assert_puzzle_hash!(P2_CONTROLLER_PUZZLE_PUZZLE => P2_CONTROLLER_PUZZLE_PUZZLE_HASH);
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_p2_controller_puzzle() -> anyhow::Result<()> {
+        let mut sim = Simulator::new();
+        let ctx = &mut SpendContext::new();
+
+        let controller_puzzle = node_from_bytes(&mut ctx.allocator, &hex!("01"))?;
+        let controller_puzzle_hash = ctx.tree_hash(controller_puzzle);
```

Comment:
You could use `1.tree_hash()` if you wanted

Reply:
let's leave it general :)